In [3]:
#include<iostream>
#include<bits/stdc++.h>
#include <cmath>
using namespace std;

In [8]:
void uninitializedVariable(){
    int a;
    cout<<"Un-initialized: will give unexpected result: "<<a;
}
uninitializedVariable();

Un-initialized: will give unexpected result: 32766

1. Constant: must be initialized

In [11]:
const double PI = 3.1415926;  // Need to initialize
cout<<PI;

3.14159

### 1. DataType:
`sizeof(type)` is used to get the size.
1. Integer:
    - **signed int:** The integers (except char) are **signed number** (which can hold zero, positive and negative numbers).
    - **unsigned int:** unsigned integers (which can hold zero and positive numbers).


In [50]:
void dataType(){
    cout<<"Integers:\n";
    cout<<"\nint: 4 bytes\n";
    cout << "sizeof(int/signed int) is " << sizeof(int) << " bytes " << " range "<< -pow(2,sizeof(int)*8-1) <<" to "<< (pow(2,sizeof(int)*8-1))-1 << endl;
    cout << "sizeof(unsigned int) is " << sizeof(unsigned int) << " bytes " << " range "<<0 <<" to "<<(pow(2,sizeof(int)*8))-1<< endl;
    
    cout<<"\nchar: 1 bytes\n";
    cout << "sizeof(char) is " << sizeof(char) << " bytes " << " range "<< -pow(2,sizeof(char)*8-1) <<" to "<<pow(2,((sizeof(char)*8)-1))-1<< endl;
    cout << "sizeof(unsigned char) is " << sizeof(unsigned char) << " bytes " << " range "<< 0 <<" to "<<pow(2,(sizeof(unsigned char)*8))-1<< endl;

    cout<<"\nshort: 2 bytes\n";
    cout << "sizeof(short/short int/signed short/ signed short int) is " << sizeof(short) << " bytes " << " range "<< -pow(2,sizeof(short)*8-1) <<" to "<< (pow(2,sizeof(short)*8-1))-1 << endl;
    cout << "sizeof(unsigned short/unsigned short int) is " << sizeof(unsigned short int) << " bytes " << " range "<< 0 <<" to "<< (pow(2,sizeof(unsigned short int)*8))-1 << endl;


    cout<<"\nLong: typical(8) bytes but minimum 4 bytes\n";
    cout << "sizeof(long/long int/signed long/ signed long int) is " << sizeof(long) << " bytes " << " range "<< -pow(2,sizeof(signed long int)*8-1) <<" to "<< (pow(2,sizeof(signed long int)*8-1))-1 << endl;
    cout << "sizeof(long/long int/signed long/ signed long int) is " << sizeof(signed long int) << " or 4 bytes " << " range "<< -pow(2,4*8-1) <<" to "<< (pow(2,4*8-1))-1 << endl;
    cout << "sizeof(unsigned long/unsigned long int) is " << sizeof(unsigned long int) << " bytes " << " range "<< 0 <<" to "<< (pow(2,sizeof(unsigned long int)*8))-1 << endl;

    cout<<"\nlong long: 8 bytes\n";
    cout << "sizeof(long long/long long int/signed long long/ signed long long int) is " << sizeof(long long) << " bytes " << " range "<< -pow(2,sizeof(signed long long int)*8-1) <<" to "<< (pow(2,sizeof(signed long long int)*8-1))-1 << endl;
    cout << "sizeof(unsigned long long/unsigned long long int) is " << sizeof(unsigned long long int) << " bytes " << " range "<< 0 <<" to "<< (pow(2,sizeof(unsigned long long int)*8))-1 << endl;
    cout<<"End of Integers:\n";

    cout<<"\nReal Number\n";
    cout<<"float 4 bytes"<<endl;
    cout << "sizeof(float) is " << sizeof(float) << " bytes " << " range "<< -pow(2,sizeof(float)*8-1) <<" to "<< (pow(2,sizeof(float)*8-1))-1 << endl;
    cout << "sizeof(double) is " << sizeof(double) << " bytes " << " range "<< -pow(2,sizeof(double)*8-1) <<" to "<< (pow(2,sizeof(double)*8-1))-1 << endl;
    cout << "sizeof(long double) is " << sizeof(long double) << " bytes " << " range "<< -pow(2,sizeof(long double)*8-1) <<" to "<< (pow(2,sizeof(long double)*8-1))-1 << endl;

    cout<<"double 8 bytes"<<endl;
    cout<<"long double 12 bytes typical(8)"<<endl;
    cout<<"\nEnd of Real Number\n";

    cout<<"\nWide Characters 2 but tipically (4) bytes\n";
    cout << "sizeof(wchar_t/char16_t/char32_t) is " << sizeof(char32_t) << " bytes " << " range "<< -pow(2,sizeof(char32_t)*8-1) <<" to "<< (pow(2,sizeof(char32_t)*8-1))-1 << endl;

   cout << "sizeof(bool) is " << sizeof(bool) << " bytes " << endl;
}
dataType();

Integers:

int: 4 bytes
sizeof(int/signed int) is 4 bytes  range -2.14748e+09 to 2.14748e+09
sizeof(unsigned int) is 4 bytes  range 0 to 4.29497e+09

char: 1 bytes
sizeof(char) is 1 bytes  range -128 to 127
sizeof(unsigned char) is 1 bytes  range 0 to 255

short: 2 bytes
sizeof(short/short int/signed short/ signed short int) is 2 bytes  range -32768 to 32767
sizeof(unsigned short/unsigned short int) is 2 bytes  range 0 to 65535

Long: typical(8) bytes but minimum 4 bytes
sizeof(long/long int/signed long/ signed long int) is 8 bytes  range -9.22337e+18 to 9.22337e+18
sizeof(long/long int/signed long/ signed long int) is 8 or 4 bytes  range -2.14748e+09 to 2.14748e+09
sizeof(unsigned long/unsigned long int) is 8 bytes  range 0 to 1.84467e+19

long long: 8 bytes
sizeof(long long/long long int/signed long long/ signed long long int) is 8 bytes  range -9.22337e+18 to 9.22337e+18
sizeof(unsigned long long/unsigned long long int) is 8 bytes  range 0 to 1.84467e+19


End of Integers:

Real Number
float 4 bytes
sizeof(float) is 4 bytes  range -2.14748e+09 to 2.14748e+09
sizeof(double) is 8 bytes  range -9.22337e+18 to 9.22337e+18
sizeof(long double) is 16 bytes  range -1.70141e+38 to 1.70141e+38
double 8 bytes
long double 12 bytes typical(8)

End of Real Number

Wide Characters 2 but tipically (4) bytes
sizeof(wchar_t/char16_t/char32_t) is 4 bytes  range -2.14748e+09 to 2.14748e+09
sizeof(bool) is 1 bytes 


1.2 ***`Header <climits>`:***
The climits header (ported to C++ from C's limits.h) contains information about limits of integer type.

In [52]:
void intTypeLimit(){
    cout << "int max = " << INT_MAX << endl;
    cout << "int min = " << INT_MIN << endl;
    cout << "unsigned int max = " << UINT_MAX << endl;
    cout << "long long max = " << LLONG_MAX << endl;
    cout << "long long min = " << LLONG_MIN << endl;
    cout << "unsigned long long max = " << ULLONG_MAX << endl;
    cout << "Bits in char = " << CHAR_BIT << endl;
    cout << "char max = " << CHAR_MAX << endl;
    cout << "char min = " << CHAR_MIN << endl;
    cout << "signed char max = " << SCHAR_MAX << endl;
    cout << "signed char min = " << SCHAR_MIN << endl;
    cout << "unsigned char max = " << UCHAR_MAX << endl;
}
intTypeLimit();

int max = 2147483647
int min = -2147483648
unsigned int max = 4294967295
long long max = 9223372036854775807
long long min = -9223372036854775808
unsigned long long max = 18446744073709551615
Bits in char = 8
char max = 127
char min = -128
signed char max = 127
signed char min = -128
unsigned char max = 255


1.3 ***`Header <cfloat>`:***
The climits header (ported to C++ from C's limits.h) contains information about limits of integer type.

In [57]:
void floatTypeLimit(){
    std::cout << "Maximum float value: " << FLT_MAX << " and " << sizeof(FLT_MAX) << " bytes " << std::endl;
    std::cout << "Minimum float value: " << FLT_MIN << " and " << sizeof(FLT_MIN) << " bytes " << std::endl;
    std::cout << "Number of decimal digits of precision: " << FLT_DIG << std::endl;
}
floatTypeLimit();

Maximum float value: 3.40282e+38 and 4 bytes 
Minimum float value: 1.17549e-38 and 4 bytes 
Number of decimal digits of precision: 6


### 2. Literal:
A literal is a specific constant value, such as 123, -456, 3.14, 'a', "Hello", that can be assigned directly to a variable; or used as part of an expression. They are called literals because they literally and explicitly identify their values.

#### 2.1 Interger Literal:
- 01234:  Octal 1234, Decimal 2322.
- 0x1abc: hexadecimal 1ABC, decimal 15274
- 0b10001001: binary (may not work in some compilers)

In [62]:
// Integer Literals
int number1 = 1234;       // Decimal
int number2 = 01234;      // Octal 1234, Decimal 2322
int number3 = 0x1abc;     // hexadecimal 1ABC, decimal 15274
int number4 = 0b10001001; // binary (may not work in some compilers)
cout<<" number1 = "<<number1<<" number2 = "<<number2<<" number3 = "<<number3<<" number4= "<<number4<<endl;

//  long literal
long number = 12345678L;     // Suffix 'L' for long
long sum = 123;              // int 123 auto-casts to long 123L
long long bigNumber = 987654321LL;  // Need suffix 'LL' for long long int
cout<<" number = "<<number<<" sum = "<<sum<<" bigNumber = "<<bigNumber<<endl;


 number1 = 1234 number2 = 668 number3 = 6844 number4= 137
 number = 12345678 sum = 123 bigNumber = 987654321


In [67]:
// float literal
float average = 55.66;      //  Need suffix 'f' for float.
float average1 = 55.66f;

#### 2.2 Charecter Literal `<ctype.h>`:
The cctype header (ported from C's ctype.h) provides functions such as isalpha(), isdigit(), isspace(), ispunct(), isalnum(), isupper(), islower() to determine the type of character; and toupper(), tolower() for case conversion.

In [73]:
#include<ctype.h>

char ch = 'A';
std::cout << "Character: " << ch << std::endl;
std::cout << "Is alphabetic? " << (isalpha(ch) ? "Yes" : "No") << std::endl;
std::cout << "Is uppercase? " << (isupper(ch) ? "Yes" : "No") << std::endl;
std::cout << "Is lowercase? " << (islower(ch) ? "Yes" : "No") << std::endl;
std::cout << "Is digit? " << (isdigit(ch) ? "Yes" : "No") << std::endl;
std::cout << "Converted to lowercase: " << static_cast<char>(tolower(ch)) << std::endl;


Character: A
Is alphabetic? Yes
Is uppercase? Yes
Is lowercase? No
Is digit? No
Converted to lowercase: a


#### 2.3 String Literals:
A String literal is composed of zero of more characters surrounded by a pair of double quotes, e.g., "Hello, world!", "The sum is ", "".

In [74]:
string directionMsg = "Turn Right";

### 3. Overflow/UnderFlow:
Overflow and underflow occur when the result of an arithmetic operation exceeds the maximum representable value or falls below the minimum representable value for a particular data type. In C++, overflowing or underflowing a variable can lead to unexpected behavior and incorrect results.

In [75]:
#include <limits>

int maxValue = std::numeric_limits<int>::max();
int minValue = std::numeric_limits<int>::min();

std::cout << "Maximum value of int: " << maxValue << std::endl;
std::cout << "Minimum value of int: " << minValue << std::endl;

int overflowResult = maxValue + 1;
int underflowResult = minValue - 1;

std::cout << "Overflow result: " << overflowResult << std::endl;
std::cout << "Underflow result: " << underflowResult << std::endl;

Maximum value of int: 2147483647
Minimum value of int: -2147483648
Overflow result: -2147483648
Underflow result: 2147483647


### 4. Operators:

#### 4.1: pre increment and post increment:

In [79]:
// pre increment and post increment
int x = 5;
cout<<x++<<endl;
cout<<x<<endl;
int y = 5;
cout<<++y<<endl;
cout<<y<<endl;

5
6
6
6


### 5. Type Casting:

In [93]:
void typecasting(){
    int sum = 0;
    double average;
    for (int number = 1; number <= 10; ++number) {
        sum += number; 
    }
    average = sum / 10;
    cout << "Average is " <<(int)average << endl;  // Average is 5
}
typecasting();

Average is 5
